homeassistant/components/rainforest_raven/config_flow.py

```diff
@@ -0,0 +1,159 @@
+                    title=user_input.get(CONF_NAME, DEFAULT_NAME),
+                    data={
+                        CONF_DEVICE: self._dev_path,
+                        CONF_MAC: [m.hex() for m in meter_macs],
+                    },
+                )
+
+        discovered_macs = [m.hex() for m in self._meter_macs]
+        schema = vol.Schema(
+            {
+                vol.Required(
+                    CONF_MAC, description={"suggested_value": ",".join(discovered_macs)}
+                ): str
+            }
+        )
+        return self.async_show_form(step_id="meters", data_schema=schema, errors=errors)
+
+    async def async_step_usb(self, discovery_info: usb.UsbServiceInfo) -> FlowResult:
+        """Handle USB Discovery."""
+        device = discovery_info.device
+        dev_path = await self.hass.async_add_executor_job(usb.get_serial_by_id, device)
+        unique_id = f"{discovery_info.vid}:{discovery_info.pid}_{discovery_info.serial_number}_{discovery_info.manufacturer}_{discovery_info.description}"
+        await self.async_set_unique_id(unique_id)
+        try:
+            await self._validate_device(dev_path)
+        except asyncio.TimeoutError:
+            return self.async_abort(reason="timeout_connect")
+        except (ParseError, SerialException):
+            return self.async_abort(reason="cannot_connect")
+        return await self.async_step_meters()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initiated by the user."""
+        if self._async_in_progress():
+            return self.async_abort(reason="already_in_progress")
+        ports = await self.hass.async_add_executor_job(serial.tools.list_ports.comports)
+        existing_devices = [
+            entry.data[CONF_DEVICE] for entry in self._async_current_entries()
+        ]
+        unused_ports = [
+            usb.human_readable_device_name(
+                port.device,
+                port.serial_number,
+                port.manufacturer,
+                port.description,
+                port.vid,
+                port.pid,
+            )
+            for port in ports
+            if port.device not in existing_devices
+        ]
+        if not unused_ports:
+            return self.async_abort(reason="no_devices_found")
+
+        errors = {}
+        if user_input is not None and user_input.get(CONF_DEVICE, "").strip():
+            port = ports[unused_ports.index(str(user_input[CONF_DEVICE]))]
+            dev_path = await self.hass.async_add_executor_job(
+                usb.get_serial_by_id, port.device
+            )
+            unique_id = _generate_unique_id(port)
+            await self.async_set_unique_id(unique_id)
+            try:
+                await self._validate_device(dev_path)
+            except asyncio.TimeoutError:
+                errors[CONF_DEVICE] = "timeout_connect"
+            except (ParseError, SerialException):
```

Comment:
Exceptions like these should be handled in the third-party library and the code that raises them.

Reply:
Can you clarify what you're looking to see here? The underlying library needs to express communication errors somehow. Would you like to see errors like this surface under a library-specific exception class?

Reply:
The library should raise an specific error for these exceptions. These exceptions are implementation details and they should be transparent for the caller.

In our [developers docs](https://developers.home-assistant.io/docs/api_lib_index) we have written the following:

> One of the foundational rules of Home Assistant is that we do not include any protocol specific code. Instead, this code should be put into a standalone Python library and published to PyPI.

Reply:
Okay, the library should only emit RAVEn-specific exceptions now.